grafana/resource_datasource_permission.go

```diff
@@ -0,0 +1,229 @@
+	if err != nil {
+		if strings.HasPrefix(err.Error(), "status: 404") {
+			log.Printf("[WARN] removing datasource permissions %d from state because it no longer exists in grafana", id)
+			d.SetId("")
+			return nil
+		}
+
+		return diag.FromErr(err)
+	}
+
+	permissionItems := make([]interface{}, len(response.Permissions))
+	count := 0
+	for _, permission := range response.Permissions {
+		permissionItem := make(map[string]interface{})
+		permissionItem["team_id"] = permission.TeamID
+		permissionItem["user_id"] = permission.UserID
+		permissionItem["permission"] = mapDatasourcePermissionInt64ToString(permission.Permission)
+
+		permissionItems[count] = permissionItem
+		count++
+	}
+
+	d.Set("permissions", permissionItems)
+
+	return nil
+}
+
+func DeleteDatasourcePermissions(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client).gapi
+
+	datasourceID := int64(d.Get("datasource_id").(int))
+
+	if err := updateDatasourcePermissions(client, datasourceID, []*gapi.DatasourcePermissionAddPayload{}, false, true); err != nil {
+		if strings.HasPrefix(err.Error(), "status: 404") {
+			log.Printf("[WARN] removing datasource permissions %d from state because it no longer exists in grafana", datasourceID)
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+func updateDatasourcePermissions(client *gapi.Client, id int64, permissions []*gapi.DatasourcePermissionAddPayload, enable, disable bool) error {
+	areEqual := func(a *gapi.DatasourcePermission, b *gapi.DatasourcePermissionAddPayload) bool {
+		return a.Permission == b.Permission && a.TeamID == b.TeamID && a.UserID == b.UserID
+	}
+
+	response, err := client.DatasourcePermissions(id)
+	if err != nil {
+		return err
+	}
+
+	if !response.Enabled {
+		if enable {
+			if err := client.EnableDatasourcePermissions(id); err != nil {
+				return err
+			}
+		} else if disable {
+			return nil
+		} else {
+			return fmt.Errorf("datasource permissions are disabled")
+		}
+	}
+
+deleteLoop:
+	for _, current := range response.Permissions {
+		for _, new := range permissions {
+			if areEqual(current, new) {
+				continue deleteLoop
+			}
+		}
+
+		err := client.RemoveDatasourcePermission(id, current.ID)
+		if err != nil {
+			return err
+		}
+	}
+
+addLoop:
+	for _, new := range permissions {
+		for _, current := range response.Permissions {
+			if areEqual(current, new) {
+				continue addLoop
+			}
+		}
+
+		err := client.AddDatasourcePermission(id, new)
+		if err != nil {
+			return err
+		}
+	}
+
+	if response.Enabled {
+		if disable {
+			if err := client.DisableDatasourcePermissions(id); err != nil {
+				return err
+			}
+		} else if enable {
+			return nil
+		} else {
+			return fmt.Errorf("datasource permissions are enabled when they should be disabled")
+		}
+	}
+
+	return nil
+}
+
+func mapDatasourcePermissionStringToInt64(permission string) int64 {
+	if permission == "Query" {
+		return 1
+	}
+	return -1
+}
```

Comment:
What happens if we map the string to `-1`? Will the API request fail? If so, maybe we should catch this earlier and let the user know the value couldn't be mapped 🤔

Reply:
Also: as in the case of the PR in the API client, if we defined this as a custom enum, then perhaps we could avoid these mappings and conversions if we define them there.

Reply:
Yeah, I made this an error. Definitely more robust. I should stop relying on the rest of the code and think a bit...

Reply:
Nah, you're doing great! <3